### PR TITLE
Offline support using service worker 

### PR DESCRIPTION
**Pull request by ZhangYiJiang**
Basic offline support via Workbox. Tested successfully on Chrome desktop.

- Fallback to `index.html` for all navigation events that generate an error 
- Caches all calls to `nusmods.com/api` - this allows the app to work offline, since `moduleList.json` and all modules on the timetable are always loaded on initial page load, this is enough data for us to render the timetable offline. 

  The caching strategy is staleWhileRevalidate, which means the request will always be fulfilled immediately from cache if available, and the cache is then refreshed in the background. This is relatively safe since most of the time the data does not change radically between visits.  

### Wishlist 

- Pre-fetch data needed for venue search in the background (module search is not very useful since the module page requires network anyway, but venue search is pretty useful) 
- Better warning/error for when the user is offline

### CODE REVIEW

File: v3/src/index.html [hunk 1]
```diff
@@ -7,6 +7,7 @@
   <link href="${htmlWebpackPlugin.files.publicPath}favicon-16x16.png" rel="icon" type="image/png" sizes="16x16">
   <link href="${htmlWebpackPlugin.files.publicPath}safari-pinned-tab.svg" rel="mask-icon" color="#ff5138">
   <link href="${htmlWebpackPlugin.files.publicPath}manifest.json" rel="manifest">
+  <link href="${htmlWebpackPlugin.options.apiBaseUrl}/moduleList.json" rel="preload">
```
Comment (li-kai) [inline v3/src/index.html]: `<link ... as="fetch">` https://developer.mozilla.org/en-US/docs/Web/HTML/Preloading_content#The_basics (scroll down a bit)

File: v3/src/js/views/errors/ErrorPage.jsx [hunk 2]
```diff
@@ -42,7 +42,7 @@ export default class NotFoundPage extends PureComponent<Props> {
             {this.errorMessage()}
           </h1>
 
-          {showRefresh &&
+          {showRefresh && navigator.onLine &&
```
Comment (li-kai) [inline v3/src/js/views/errors/ErrorPage.jsx]: We shouldn't show the report dialog either right?

File: v3/src/js/reducers/app.js [hunk 3]
```diff
@@ -14,16 +14,18 @@ import {
...
+import { SET_ONLINE_STATUS } from 'actions/online';
 
-const defaultAppState: AppState = {
+const defaultAppState = (): AppState => ({
   // Default to the current semester from config.
   activeSemester: config.semester,
   // The lesson being modified on the timetable.
   activeLesson: null,
-};
+  isOnline: navigator.onLine,
```
Comment (li-kai) [inline v3/src/js/reducers/app.js]: I think we can assume the user is online when loading the app?

Comment (ZhangYiJiang) [reply v3/src/js/reducers/app.js]: We can't, actually, because the entire app can now be run off service worker offline

File: v3/src/index.html [hunk 4]
```diff
@@ -1,24 +1,26 @@
...
-  <meta name="og:image" content="${htmlWebpackPlugin.files.publicPath}share.png">
-  <meta name="og:site_name" content="NUSMods">
-  <meta name="og:type" content="website">
-  <!-- Open Search -->
-  <link href="${htmlWebpackPlugin.files.publicPath}opensearch.xml" rel="search" type="application/opensearchdescription+xml">
-</head>
-<body>
-  <div id="app"></div>
-</body>
+  <head>
```
Comment (li-kai) [inline v3/src/index.html]: Wow can't believe none of us caught this before

**Comment by ZhangYiJiang** [comment]
What works: 

- Viewing/changing lessons on the timetable when offline 
- Viewing added modules when offline 

Can be improved 

- Adding/removing modules - adding any module that has not been previously added/visited will result in an uncaught error - this can be improved on after #484 is merged 
- Venue info should be prefetched somehow 
- Visiting module info page of any module while offline will result in an infinite loading spinner - #521 is probably needed for this

**Pull request merged**